extensions/kinesis/src/main/java/com/hazelcast/jet/kinesis/impl/KinesisSinkP.java

```diff
@@ -0,0 +1,354 @@
+                    "Sending them will be retried, but reordering might be unavoidable.");
+            List<PutRecordsResultEntry> resultEntries = result.getRecords();
+            for (int i = resultEntries.size() - 1; i >= 0; i--) {
+                PutRecordsResultEntry resultEntry = resultEntries.get(i);
+                if (resultEntry.getErrorCode() == null) {
+                    buffer.remove(i);
+                }
+            }
+        } else {
+            buffer.clear();
+        }
+    }
+
+    private enum State {
+        /**
+         * Ready to send data to Kinesis, if available.
+         */
+        READY_TO_SEND,
+
+        /**
+         * Data has been sent to Kinesis, waiting for a reply.
+         */
+        SENDING_IN_PROGRESS,
+    }
+
+    private static class Buffer<T> {
+
+        private final FunctionEx<T, String> keyFn;
+        private final FunctionEx<T, byte[]> valueFn;
+
+        private final BufferEntry[] entries;
+        private int entryCount;
+        private int totalEntrySize;
+
+        Buffer(FunctionEx<T, String> keyFn, FunctionEx<T, byte[]> valueFn) {
+            this.keyFn = keyFn;
+            this.valueFn = valueFn;
+            this.entries = initEntries();
+        }
+
+        boolean add(T item) {
+            if (entryCount == entries.length) {
+                return false;
+            }
+
+            String key = keyFn.apply(item);
+            int unicodeCharsInKey = key.length();
+            if (unicodeCharsInKey > MAX_UNICODE_CHARS_IN_KEY) {
+                throw new IllegalArgumentException("Key of " + item + " too long");
+            }
+            int keyLength = getKeyLength(key);
+
+            byte[] value = valueFn.apply(item);
+            int itemLength = value.length + keyLength;
+            if (itemLength > MAX_RECORD_SIZE_IN_BYTES) {
+                throw new IllegalArgumentException("Item " + item + " encoded length (key + payload) is too big");
+            }
+
+            if (totalEntrySize + itemLength > MAX_REQUEST_SIZE_IN_BYTES) {
+                return false;
+            } else {
+                totalEntrySize += itemLength;
+
+                BufferEntry entry = entries[entryCount++];
+                entry.set(key, value, itemLength);
+
+                return true;
+            }
+        }
+
+        public void remove(int index) { //todo: test it, at least manually
+            if (index < 0 || index >= entryCount) {
+                throw new IllegalArgumentException("Index needs to be between 0 and " + entryCount);
+            }
+
+            totalEntrySize -= entries[index].encodedSize;
+
+            int lastIndex = entryCount - 1;
+            if (index < lastIndex) {
+                BufferEntry tmp = entries[index];
+                entries[index] = entries[lastIndex];
+                entries[lastIndex] = tmp;
+            } else {
+                entryCount--;
+            }
```

Comment:
If not deleting from the end, the old code doesn't move entries, but just swaps the removed and last entry, which is wrong. It also doesn't decrement `entryCount`
```suggestion
            if (index < 0 || index >= entryCount) {
                throw new IndexOutOfBoundsException("index=" + index + ", size=" + entryCount);
            }
            totalEntrySize -= entries[index].encodedSize;
            entryCount--;
            if (index < entryCount) {
                BufferEntry tmp = entries[index];
                System.arraycopy(entries, index + 1, entries, index, entryCount - index);
                entries[entryCount] = tmp;
            }
```